From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Sun, 4 Jan 2026 13:43:12 -0300
Subject: Allow disabling GetModuleHandle16 calls

Some games, like Monster Truck Madness 2, attempt to use 16 bit handles even though they don't even need it, and for some reason that causes huge slowdowns during loading screens
---
 dlls/krnl386.exe16/ne_module.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/krnl386.exe16/ne_module.c b/dlls/krnl386.exe16/ne_module.c
index aeb0a2e3189..8cf672c6ac0 100644
--- a/dlls/krnl386.exe16/ne_module.c
+++ b/dlls/krnl386.exe16/ne_module.c
@@ -1377,6 +1377,15 @@ void WINAPI FreeLibrary16( HINSTANCE16 handle )
  */
 HMODULE16 WINAPI GetModuleHandle16( LPCSTR name )
 {
+    static int disable_win16_module_handles = -1;
+    if (disable_win16_module_handles == -1)
+    {
+        const char *val = getenv("SCRIBBLE_DISABLE_WIN16_MODULE_HANDLES");
+        disable_win16_module_handles = (val && !strcmp(val, "1"));
+    }
+    if (disable_win16_module_handles)
+        return 0;
+
     HMODULE16	hModule;
     LPSTR	s;
     BYTE	len, *name_table;
-- 
2.52.0

